### PR TITLE
Sot 31 why and how zk snark works

### PR DESCRIPTION
**Pull request by Draply**
- [x] Did you check your grammar / reader proof your document using ChatGPT?
- [x] Don't forget to squash commits into meaningful chunks before merging
- [x] After merging this, please update your task status!

### CODE REVIEW

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
+# Chapter 3
+
+## 3.1. **The Fundamental Theorem of Algebra**
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:3]: We already had the definition for this

File: articles/why-and-how-zk-snark-works/1.introduction.md
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/1.introduction.md:1]: Should have a "main" markdown that link to others (see proofs-argument....md)

File: articles/why-and-how-zk-snark-works/2.the_medium_of_a_proof.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+## The medium of a Proof.
+Let us consider **polynomials** with degree 3, which can be visualized as a curve on a graph, shaped by mathematical equations: 
+
+![](attachments/polynomial.png) 
+
+**Polynomials** have an advantageous property, if we have two non-equal **polynomials** of degree at most d, they can intersect at no more than d points. For example :
+
+![](attachments/twocrosspoly.png) 
+
+**The Fundamental Theorem of Algebra** tells us that a degree d **polynomial** can have at most d solutions , and therefore at most d shared points.
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/2.the_medium_of_a_proof.md:11]: We already had our definition for this. (Update it instead)

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md (hunk 1 above)
Comment (sonntuet1997) [reply articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:3]: Fundamental Theorem in both chapter 2 and 3 ?

File: articles/why-and-how-zk-snark-works/2.the_medium_of_a_proof.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+In fact, out of all choices of *x* to evaluate, only at most 3 choices will have equal evaluations in example polynomials above, and all others will differ. 
+So we have a simple protocol here :
+* The verifier chooses a random value for x and evaluates his polynomial locally 
+* The verifier gives *x* to the prover and asks to evaluate the polynomial in question 
+* The prover evaluates his polynomial at *x* and gives the result to the verifier 
+* The verifier checks if the local result is equal to the prover’s result, and if so then the statement is proven with a high confidence
+
+ Consider an integer range of *x* from 1 to $10^{77}$, the number of points where evaluations are different is *$10^{77}-d$*. Henceforth the probability that *x* accidentally “hits” any of the d shared points is equal to $\frac{d}{10^{77}}$ , which is considered negligible.
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/2.the_medium_of_a_proof.md:20]: Math errors

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+For example, we have:
+$$p(x) = x^{3} - 3x^{2} + 2x$$
+
+In this case, with coefficients 1, -3, 2, the encrypted polynomial can be evaluated as follows:
+
+$$
+\begin{aligned}
+E(x^{3})^{1}E(x^{2})^{-3}E(x)^{2}= \\
+ (g^{x^{3}})^{1}(g^{x^{2}})^{-3}(g^{x})^{2}= \\
+ g^{1x^{3}-3x^{x^2}g^{2x}}= \\
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:42]: 3x^x^2?

Comment (sonntuet1997) [reply articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:42]: Fix math equations

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+The Prover calculates $h(x) = p(x)$, $t(x) = x$, evaluates $$p = p(23) = 10626$$ $$h = h(23) = 23$$ and provides p, h to the Verifier.
+
+The Verifier then checks that:  $$p = t \cdot h: 10626 = 462 \cdot 23$$ 
+This is true, and therefore the statement is proven.
+
+## 3.2. Modular Arithmetic 
+
+## 3.3. Encrypted Polynomial
+
+Let's define *E* : $$E(x) = g^{v} \pmod{n}$$
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:31]: E(x) = g^v ?

File: articles/why-and-how-zk-snark-works/2.the_medium_of_a_proof.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# Chapter 2 
+## The medium of a Proof.
+[Fundamental theorem of Algebra](./facts/fundamental_theorem_of_algebra.md)
```
Comment (ktokto313) [inline articles/why-and-how-zk-snark-works/2.the_medium_of_a_proof.md:3]: link doesn't work

Comment (Draply) [reply articles/why-and-how-zk-snark-works/2.the_medium_of_a_proof.md:3]: i will fix when sot-40's merged into main

**Review by haipro287 — CHANGES_REQUESTED** [review]

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+* The Verifier could collude with the Prover and disclose those secret parameters s, α.
+* The Verifier has to store $\alpha$ and $t(s)$ until all relevant proofs are verified.
+
+Hence, we need to
+
+ secure the secrets $(t(s),\alpha)$, and this is where *cryptographic pairing* fits in.
+
+### 3.6.1. Multiplication of Encrypted Values
+
+Cryptographic pairings (bilinear map) have the core properties that can be expressed as:
```
[the remark this answers is not in the document]
Comment (Draply) [reply articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:123]: We have term for this, so i will link to it, when it merged

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+*Verifier*:
+
+* Check that $p=t(s) \cdot h$, meaning: 
+  $g^{p} = (g^{h})^{t(s)}$
+
+If the Prover claims to have a satisfactory polynomial using only 2 powers $s^{3}$ and $s^{1}$, it is not possible to verify in the current protocol.
+
+## 3.4. Restricting a Polynomial
+
+The main idea is to sample a random $\alpha$, and the provider needs to calculate $a^{,} = a^{\alpha} \pmod{n}$, then provide the tuple $(a,a^{,})$ using an elliptic curve. This ensures that the provider can't extract $\alpha$ from the tuple $(a,a^{,})$ through brute force.
```
[the remark this answers is not in the document]
Comment (Draply) [reply articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:70]: I wrote the problem in the foot of 3.3

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+*Prover*:
+
+* Calculates $h(x) = \frac{p(x)}{t(x)}$
+* Evaluates: $E(p(s))$, $E(h(s))=g^{h(s)}$
+* Provides $g^{p}$, $g^{h}$
+
+*Verifier*:
+
+* Check that $p=t(s) \cdot h$, meaning:
```
[the remark this answers is not in the document]
Comment (Draply) [reply articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:63]: sorry, i will add in Notes.md

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+The Verifier samples a random value 23, calculates  $$t = t(23) = (23 - 1)(23 - 2) = 462$$ and gives 23 to the Prover.
+
+The Prover calculates $h(x) = p(x)$, $t(x) = x$, evaluates $$p = p(23) = 10626$$ $$h = h(23) = 23$$ and provides p, h to the Verifier.
+
+The Verifier then checks that:  $$p = t \cdot h: 10626 = 462 \cdot 23$$ 
+This is true, and therefore the statement is proven.
+
+## 3.2. Modular Arithmetic 
+
+## 3.3. Encrypted Polynomial
```
[the remark this answers is not in the document]
Comment (Draply) [reply articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:29]: sorry, i will add this missing parts into term and link them

**Review by haipro287 — COMMENTED** [review]
the "why" part is still missing, your summary only have the "how" part

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md (hunk 5 above)
Comment (haipro287) [reply articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:32]: @Draply resolved?

**Review by haipro287 — DISMISSED** [review]

**Review by haipro287 — COMMENTED** [review]
look not good to me :v

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md (hunk 5 above)
Comment (Draply) [reply articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:37]: oh! i forgot to commit it . *v* is monomial so it has x in

**Review by haipro287 — APPROVED** [review]

File: articles/why-and-how-zk-snark-works/1.introduction.md
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/1.introduction.md:1]: The file name should be `1_introduction.md` instead

Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/1.introduction.md:1]: rename the pdf file to something more meaningful

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+In particular:
+Given:
+$$p(x)  = x^3 - 3x^2 + 2x$$
+
+$$t(x) = (x-1)(x-2)$$
+
+The Verifier samples a random value 23, calculates  $$t = t(23) = (23 - 1)(23 - 2) = 462$$ and gives 23 to the Prover.
+
+The Prover calculates $h(x) = p(x)$, $t(x) = x$, evaluates $$p = p(23) = 10626$$ $$h = h(23) = 23$$ and provides p, h to the Verifier.
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:27]: p, h should be placed inside `$$`

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+
+The Verifier then checks that:  $$p = t \cdot h: 10626 = 462 \cdot 23$$ 
+This is true, and therefore the statement is proven.
+
+## 3.2. Modular Arithmetic 
+
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:33]: Nothing here?

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+*Verifier*:
+
+* Generate a random *s*
+* Calculate $E(s^{i}) = g^{s^{i}}$: same as the example above
+* Evaluate the *target polynomial* with *s*: *t(s)*
+* Provide to the Prover: $E(s^{0}),E(s^{1}),E(s^{2}),...,E(s^{d})$
+
+*Prover*:
+
+* Calculates $h(x) = \frac{p(x)}{t(x)}$
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:63]: should be h(s) = p(s)\t(s)

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+*Verifier*:
+
+* Check that $p=t(s) \cdot h$, meaning: 
+  $$g^{p} = (g^{h})^{t(s)}$$
+
+If the Prover claims to have a satisfactory polynomial using only 2 powers $s^{3}$ and $s^{1}$, it is not possible to verify in the current protocol.
+
+## 3.4. Restricting a Polynomial
+
+The main idea to resolve above problem is to sample a random $\alpha$, and the provider needs to calculate $a^{,} = a^{\alpha} \pmod{n}$, then provide the tuple $(a,a^{,})$ using an elliptic curve. This ensures that the provider can't extract $\alpha$ from the tuple $(a,a^{,})$ through brute force.
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:76]: knowledge of exponent?

File: articles/why-and-how-zk-snark-works/Notes.md
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/Notes.md:1]: the file name should be `notes.md`

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (sonntuet1997) [reply articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:76]: Given infinite power, someone could actually brute force this.
-> The "brute force" might cause some confusion

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+In summary:
+
+*Verifier*:
+
+* Chooses 2 random *s*, $\alpha$ and provides evaluations for $x=s$ and $(g^{s},g^{\alpha \cdot s})$
+* *Prover*: Applies the coefficient c:  $((g^{s})^c,(g^{\alpha \cdot s})^{c})=(g^{c \cdot s},g^{\alpha \cdot s \cdot c})$
+* *Verifier*: Checks
+
+In particular, for a degree $d$ polynomial:
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:90]: degree d but the formulas is x^{3} - 3x^{2} + 2x?

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+*Prover*:
+
+* Evaluates the encrypted polynomial with $s$: 
+$$g^{p} = g^{p(s)} =(g^{s_{0}})^{c_{0}} \cdot (g^{s_{1}})^{c_{1}} \cdots (g^{s_{d}})^{c_{d}} = g^{c_{0}s^{0} +c_{1}s^{1}+ \cdots+c_{d}s^{d}  } =g^{s^{3}} \cdot g^{-3s^{2}} \cdot g^{2s}=g^{s^{3}-3s^{2}+2s}$$       
+$$\Rightarrow g^{{p}'} = g^{\alpha (s^{3}-3s^{2}+2s)}$$ 
+*Verifier* checks :
+  $$(g^{p})^{\alpha}=g^{{p}'}$$ 
+  $$\Rightarrow (g^{s^{3}-3s^{2}+2s})^{\alpha} = g^{\alpha (s^{3}-3s^{2}+2s)}$$ 
+
+##### Conclusion
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:106]: Should be placed inside 

> [!NOTE]
> asd 
> sadsad

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+
+* We can be sure that the Prover did not use anything else other than the polynomial provided by the Verifier since there is no other way to preserve $\alpha$. 
+* However, there is still a significant drawback: *The Verifier could **brute-force** a limited range of coefficient combinations until the result is **equal** to the Prover's answer*.
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:109]: What is the purpose of the verifier doing the brute force?

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+## 3.6. Non-Interactivity 
+
+Up to this point, the proof is only valid for the original Verifier because:
+
+* The Verifier could collude with the Prover and disclose those secret parameters s, α.
+* The Verifier has to store $\alpha$ and $t(s)$ until all relevant proofs are verified.
+
+Hence, we need to
+
+ secure the secrets $(t(s),\alpha)$, and this is where *cryptographic pairing* fits in.
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:125]: link to the pairing

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+
+### 3.6.1. Multiplication of Encrypted Values
+
+Cryptographic pairings (bilinear map) have the core properties that can be expressed as:
+$$e(g^{a},g^{b})=e(g^{b},g^{a})=e(g^{ab},g^{1})=e(g^{1},g^{ab})=e(g^{1},g^{a})^{b}=e(g^{1},g^{1})^{ab}= \cdots$$
+
+### 3.6.2. Trusted Party Setup 
+
+We trust a single party to generate secrets $s$ and $\alpha$. As soon as $\alpha$ and all necessary powers of $s$ with *shift* are encrypted, they must be deleted. So we use a *common reference string* or **CRS** to generate these parameters.
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:134]: CRS should be a separate term

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+Hence, we need to
+
+ secure the secrets $(t(s),\alpha)$, and this is where *cryptographic pairing* fits in.
+
+### 3.6.1. Multiplication of Encrypted Values
+
+Cryptographic pairings (bilinear map) have the core properties that can be expressed as:
+$$e(g^{a},g^{b})=e(g^{b},g^{a})=e(g^{ab},g^{1})=e(g^{1},g^{ab})=e(g^{1},g^{a})^{b}=e(g^{1},g^{1})^{ab}= \cdots$$
+
+### 3.6.2. Trusted Party Setup
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:132]: Link

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+### 3.6.3. Trusting One out of Many 
+
+There is no way to prove that $\alpha$ and s are deleted. So we can generate a composite CRS by multiple parties. 
+
+In particular: 
+* Alice samples her random $s_{A}$ and $\alpha_{A}$ and publishes her CRS:  $g^{s_{A}^{i}},g^{\alpha_{A}},g^{\alpha_{A}s_{A}^{i}}$  
+* Bob samples his $s_{B}$ and $\alpha_{B}$ and  Alice's CRS through homomorphic multiplication: $(g^{s_{A}^{i}})^{s_{B}^{i}},(g^{\alpha_{A}})^{\alpha_{B}},(g^{\alpha_{A}s_{A}^{i}})^{\alpha_{B}s_{B}^{i}}$, then publishes A-B's CRS: $(g^{s_{AB}^{i}},g^{\alpha A B},g^{\alpha_{AB}s_{AB}^{i}})$ 
+* So does with C, D, E, $\cdots$.
+
+**Conclusion**: *The more there are unrelated participants in CRS setup, the fainter the possibility of fake proofs*
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:145]: Should be in:
> [!NOTE]
> sds

File: articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+
+* Calculate $h(x)= \frac{p(x)}{t(x)}$ 
+* Evaluate $g^{p(s)}$ and $g^{h(s)}$ using $g^{s^{i}}$ 
+* Evaluate $g^{\alpha p(s)}$ using $g^{\alpha s^{i}}$ 
+* Sample random $\delta$ 
+* Set the randomized proof $\pi = (g^{\delta p(s)}g^{\delta h(s)},g^{\delta \alpha p(s)})$
+
+#### Verification 
+Check:
+* $e({g^{{p}'}},g)=e(g^{p},g^{\alpha})$
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/3.non_interactive_zero-knowledge_of_a_polynomial.md:170]: where is $\pi$ ?

File: articles/why-and-how-zk-snark-works/4.general_purpose_zero_knowledge_proofs.md [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+	if w then 
+		return a x b
+	else 
+		return a + b
+	else if 
+end function
+```
+
+assuming $\omega \in \{0,1\}$  : 
+$f(\omega,a,b) = \omega (a * b ) + (1- \omega )(a*b)$
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/4.general_purpose_zero_knowledge_proofs.md:21]: wrong

File: articles/why-and-how-zk-snark-works/4.general_purpose_zero_knowledge_proofs.md [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+```
+
+assuming $\omega \in \{0,1\}$  : 
+$f(\omega,a,b) = \omega (a * b ) + (1- \omega )(a*b)$  
+## 4.2. Single Operation 
+Any computation at it is core consists of elemental operations of the form:
+                 *left operand*  **operator** *right operand* = *output*
+### 4.2.1. Arithmetic Properties of Polynomials
+Let us consider two following polynomials : 
+$f(x) = 2x^{2}-9x+10$
```
Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/4.general_purpose_zero_knowledge_proofs.md:27]: Fix the format

Comment (sonntuet1997) [inline articles/why-and-how-zk-snark-works/4.general_purpose_zero_knowledge_proofs.md:27]: Screenshot 2024-01-09 at 00 17 20
Fix the format

**Review by ktokto313 — APPROVED** [review]

**Review by Tranduy1dol — APPROVED** [review]

**Pull request closed without merging**